### DISCUSSION TITLE
How to handle multiple channels

### DISCUSSION THREAD

**Discussion by bnjjj** [libp2p/rust-libp2p#2291] [Q&A]
Hello, I will explain my need, indeed I'm trying to move from a homemade solution with tokio TcpStream to a libp2p implementation. For now I have 2 opened tcp stream in my solution with 2 different ports. One to send specific events and another one to send commands. I'm wondering what is the best way to handle this kind of multiple channels thanks to libp2p. I saw yamux, I saw mplex, I read about multiplexing but I'm a little bit lost. Knowing that each nodes could send or receive data on these 2 channels.

Thanks for your help

**Comment by thomaseizinger** [comment]
For libp2p you will want to think in _protocols_.

As an intermediary transition, you can make two protocols (implying two NetworkBehaviours in the end), one for events and one for commands.

Likely, as you learn more about libp2p, you may find that the network communication of your app can be split into more individual domains, all representing an individual protocol.

You could also make a single NetworkBehaviour for each individual message (ie each event and each command). They could all be modeled via the `RequestResponse` behaviour (with an empty response perhaps).

**Reply by bnjjj** [reply]
Thanks for your answer. So at the end I will have 2 swarm instances right ?

**Reply by thomaseizinger** [reply]
> Thanks for your answer. So at the end I will have 2 swarm instances right ?

Two Swarms are not necessary as there is a custom derive for `NetworkBehaviour` that allows you to compose multiple Behaviours into a tree structure so you only pass one into the Swarm.

Checkout some of the examples :)

**Comment by bnjjj** [comment]
I have a related question. I used different topic to send Event or Message and keep the same kind of behavior. But I'm wondering if topics share the same connection ? I mean if I send a lot of events on the topic Event will I have some latencies on the Message topic ?

**Reply by thomaseizinger** [reply]
With `Topic` you are saying you are using `Gossipsub`?

If you don't have/want a mesh network of multiple peers, using `Gossipsub` seems unnecessary. From what I understand so far, using `RequestResponse` is probably the best fit.

I don't fully understand the current design you went with so I cannot answer your latency question, do you mind sharing some code?

**Reply by bnjjj** [reply]
For now I only use the example `chat-tokio` with floodsub and add handling of topic inside floodsub. Here is an example:

```rust
impl NetworkBehaviourEventProcess<FloodsubEvent> for MyBehaviour {
        // Called when `floodsub` produces an event.
        fn inject_event(&mut self, message: FloodsubEvent) {
            let floodsub_topic: floodsub::Topic = floodsub::Topic::new("chat");
            let floodsub_event_topic: floodsub::Topic = floodsub::Topic::new("event");
            if let FloodsubEvent::Message(message) = message {
                if message.topics.contains(&floodsub_topic) {
                    println!(
                        "Message Received: '{:?}' from {:?}",
                        String::from_utf8_lossy(&message.data),
                        message.source
                    );
                }
                if message.topics.contains(&floodsub_event_topic) {
                    println!(
                        "Event Received: '{:?}' from {:?}",
                        String::from_utf8_lossy(&message.data),
                        message.source,
                    );
                }
            }
        }
    }
```

**Reply by thomaseizinger** [reply]
Like I said, I don't think Floodsub / Gossipsub is a good fit for your problem. Have you tried using the `RequestResponse` `NetworkBehaviour`?
